cmd/controller/apps_reconciler.go

```diff
@@ -35,5 +43,53 @@ func (r *AppsReconciler) Reconcile(request reconcile.Request) (reconcile.Result,
 		return reconcile.Result{}, err
 	}
 
-	return r.appFactory.NewCRDApp(existingApp, log).Reconcile()
+	force := false
+	crdApp := r.appFactory.NewCRDApp(existingApp, log)
+	if !r.areAppRefsUpToDate(crdApp.GetSecretRefs(), secret, existingApp) ||
+		!r.areAppRefsUpToDate(crdApp.GetConfigMapRefs(), configmap, existingApp) {
+		force = true
+	}
+
+	result, err := crdApp.Reconcile(force)
+	if err != nil {
+		return result, err
+	}
+
+	// Only update if reconcile is successful.
+	// Leave as not updated if error occurs
+	// in case follow up reconcile request may
+	// address issue for App.
+	r.appRefTracker.MarkAppUpdated(existingApp.Name, existingApp.Namespace)
+
+	return result, err
+}
+
+func (r *AppsReconciler) areAppRefsUpToDate(refNames map[string]struct{}, kind string, app *v1alpha1.App) bool {
+	// No updates if paused or cancelled
+	if app.Spec.Canceled || app.Spec.Paused {
+		return true
+	}
+
+	// If App is being deleted, remove App from appRefTracker.
+	if app.DeletionTimestamp != nil {
+		for refName := range refNames {
+			r.appRefTracker.RemoveAppFromRefMap(kind, refName, app.Namespace, app.Name)
+			r.appRefTracker.RemoveAppFromUpdateMap(app.Name, app.Namespace)
+		}
+		return true
+	}
+
+	// Make sure refs for App are always up to date
+	// in appRefTracker.
+	for refName := range refNames {
+		if !r.appRefTracker.CheckAppExistsForRef(kind, refName, app.Namespace, app.Name) {
+			r.appRefTracker.AddAppToRefMap(kind, refName, app.Namespace, app.Name)
+		}
+	}
+
```

Comment:
this method is called areAppRefsUpToDate(). to me that reads that this method just checks on something. in highlighted code, it's doing some modification which i think is surprising for a checking method. is there a better place to put it (at the least separate function).

Reply:
Sure, there can be an `updateAppRefs()` method that runs every time with these aspects of updating the AppRefTracker and then the check method of `appRefTracker.GetAppUpdateStatus(app.Name, app.Namespace)` can be used directly to determine whether to set force.

Reply:
Updated func name and have it solely focus on updating app refs.